### DISCUSSION TITLE
Prevent account enumeration

### DISCUSSION THREAD

**Discussion by davidpiendl** [keycloak/keycloak#15632] [Q&A]
Hi together,

we recently found out that creating a token with client credentials via oidc leads to different responses if the client does not exist or if the secret is invalid (Statuscode 400 vs 401). Is there a way to answer the token request with a more generic response?

We are currently using v19.0.3-legacy in our environment. I found this solved ticket https://github.com/keycloak/keycloak/issues/12162 but is it possible, that its only available in the Quarkus distribution of keycloak?

Thank you and best regards
David

**Comment by darius-m** [comment]
The title mentions account enumeration, but the text mentions client enumeration. Please clarify which is the correct question.

For the client names, the information is basically public since the name is included in the redirect URL when the client sends a request to the authentication server (Keycloak). Because of this, a possible attacker can then fetch the corresponding name if they know any of the clients that use Keycloak for authentication.

If the attacker can simply guess a valid name, in theory they should not be able to find a valid redirect URL without prior knowledge, since Keycloak validates the redirect URL that is sent with the authentication request, and returns an `invalid request URI` (or similar) response (admittedly, I am not sure whether an attacker could use Keycloak responses to guess a valid redirect URI).

Ultimately, if the attacker knows both the client name and a valid redirect URI (this should be easiest if they already know one of the clients), most users should be protected by the URI verification, unless their system is compromised in a way that would let an attacker fake the client site (since most sites should use HTTPS, this would probably require having the user trust a spoofed or invalid certificate).

**Reply by davidpiendl** [reply]
Thank you for your detailed explanation. I am not sure if i explained it correctly. 

In our case we saw the different responses on the token endpoint https://-keycloakurl-/auth/realms/-realmname-/protocol/openid-connect/token. If a client creates an access token with client_credentials via this endpoint, the responses are different for unknown client-id (400) and known client-id with invalid secret (401). As far as I know, it should be a generic response in both cases, so a possible attacker can not gather information out of the response he gets.

I hope its now a bit clearer.

**Reply by darius-m** [reply]
Ah, so you mean that the client's credentials can be brute forced based on the response code. This can indeed be a case of concern, I don't think it can fall under the "enumeration" category of attacks.

However, since the client's name is public information (as long as an attacker knows any one of the services that you have secured using Keycloak), the brute force attack would likely be able to simply rely on a valid response from the server in case the credentials are correct. And, unless your clients are all private (behind a secured proxy or VPN), an attacker can reasonably easily find a public URL that redirects users to Keycloak for login, in which case the client's name is in the Keycloak URL - e.g. `[...]/protocol/openid-connect/auth?client_id=account-console` is part of the URL when you log in from the user account console.

**Reply by sschu** [reply]
For client credentials grants, the client name is normally not publicly known. It is used for things like machine-to-machine communication.

**Reply by darius-m** [reply]
Ah, sorry, I was not entirely familiar with the name, it appears. Yes, for client credentials grant (which would mean a service account) that would indeed be the case if the client is not used for user authentication. I did not properly consider this case.

As a temporary solution, if an upgrade to KC20 is not possible, the status code may be intercepted by a reverse proxy, so 400 is returned to the client if the KC backend returns 401.

**Comment by sschu** [comment]
@davidpiendl I think the fix is only available in KC20, not in KC19.

**Reply by davidpiendl** [reply]
Thank you for the information.
